packages/react-native-web/src/exports/Image/index.js

```diff
@@ -332,24 +340,94 @@ const Image: React.AbstractComponent<
   );
 });
 
-Image.displayName = 'Image';
+/**
+ * This component handles specifically loading an image source with header
+ */
+const ImageWithHeaders: ImageComponent = React.forwardRef((props, ref) => {
+  // $FlowIgnore
+  const nextSource: SourceObject = props.source;
+  const prevSource = React.useRef<SourceObject>({});
+  const cleanup = React.useRef<Function>(() => {});
+  const [blobUri, setBlobUri] = React.useState('');
+
+  const { onError, onLoadStart } = props;
+
+  React.useEffect(() => {
+    if (!hasSourceDiff(nextSource, prevSource.current)) return;
+
+    // When source changes we want to clean up any old/running requests
+    cleanup.current();
+
+    prevSource.current = nextSource;
+
+    let uri;
+    const abortCtrl = new AbortController();
+    const request = new Request(nextSource.uri, {
+      headers: nextSource.headers,
+      signal: abortCtrl.signal
+    });
+    request.headers.append('accept', 'image/*');
+
+    if (onLoadStart) onLoadStart();
+
+    fetch(request)
+      .then((response) => response.blob())
+      .then((blob) => {
+        uri = URL.createObjectURL(blob);
+        setBlobUri(uri);
+      })
+      .catch((error) => {
+        if (error.name !== 'AbortError' && onError) {
+          onError({ nativeEvent: error.message });
+        }
+      });
```

Comment:
I've tried to write some unit tests covering the new functionality, but because of this logic a few mocks need to be setup and make the PR bigger than it has to be 

I think it might be better to move this logic to `ImageLoader.loadUsingHeaders` so in unit tests we can stub that method and ensure it gets called when we intent to 

It's easy to manually verify the `fetch` logic works and downloads content 
Once that's clear it's easy to just verify `loadUsingHeaders` is called correctly with expected input

---

I've made a similar comment on the main repo: https://github.com/necolas/react-native-web/pull/2442/files#r1072080652

--- 

What do you think, should we refactor / write some tests now or let the mainstream repo request those changes?

Reply:
Hmm since this logic is the real meat & potatoes of the `ImageWithHeaders` component, if we move it to `ImageLoader.loadUsingHeaders` I assume we could get rid of this `ImageWithHeaders` component, and basically check if there's headers passed in the props - if yes, we call `ImageLoader.loadUsingHeaders` instead of `ImageLoader.load` here?

https://github.com/Expensify/react-native-web/blob/bd1f7b887ad690ef926f0c3d9f97c1f625d4bd55/packages/react-native-web/src/exports/Image/index.js#L269-L294

If that's what you're thinking, I honestly do think that logic is cleaner, even without needing to write tests for Expensify's case so I'd say the refactor sounds like a nice idea

Reply:
Well, the idea is purely to ease mocking but I'll give it a try

Last PR tried to solve everything inside the same component but that resulted in more logic

What a component like `ImageWithHeaders` gives us is a guarantee that `source` would always be an object with headers

BTW there's feedback on the mainstream PR: https://github.com/necolas/react-native-web/pull/2442#issuecomment-1396183044 that we should write tests and thumbs up for extracting `ImageLoader.loadUsingHeaders`

Reply:
I've tried to remove the `ImageWithHeaders` component and use either `loadUsingHeaders` or `load` functions here: https://github.com/Expensify/react-native-web/compare/master...kidroca:react-native-web:kidroca/feat/image-loader-headers-alt-2?diff=unified

But it results in a similar amount of changes and modifies some of the original logic (and seems harder to review)
- because now the source loading `useEffect` needs to account for objects
- `load` and `loadWithHeaders` need to work in a similar way in order to be interchangeable (so they were modified to return a `request.cancel` function)

We might merge `load` and `loadUsingHeaders` instead of testing which one to run, though this would be similar to the first PR were `load` handled loading images with and without headers

Reply:
Thanks for doing that refactor! Personally I prefer the new changes because it keeps all the image loading logic in `ImageLoader` and there's minimal changes to `Image/index.js` - I don't see those new changes too difficult to review (though I am not sure where `lastLoadedSource` gets updated).

I'd say let's move forward with this last refactor, as I think it's pretty straightforward compared to passing / not passing specific props to the BaseImage component required in this PR

Reply:
IMO the refactor is more of a proof that there are more "gymnastics" necessary in order to make this work by changing the original logic inside Image component 

Original logic is changed, the cleanup logic is different, `ImageLoader.load` is changed, people, including the mainstream maintainer would have to inspect how these used to work and whether the change is suitable 

IMO the mainstream maintainer already saw the update and is fine with just moving the `fetch` call to `ImageLoader.loadUsingHeaders`. I'm still trying different things, but the most straightforward PR so far seems to be the current one

There's also a big rework planned for the Image and the original loading logic would change, it would probably be best to not write stuff that depend on it (`loadUsingHeaders` depends on `load`)

Reply:
Let me update the current PR with `loadUsingHeaders` extracted to `ImageLoader` and then we can make one final decision  which set of changes would work best for us

Reply:
> IMO the mainstream maintainer already saw the update and is fine with just moving the fetch call to `ImageLoader.loadUsingHeaders`. I'm still trying different things, but the most straightforward PR so far seems to be the current one

Yeah this is one additional reason I really like this approach, even though there may be some additional changes in the upstream repo needed that we don't need at the moment in this fork 👍 

> Let me update the current PR with `loadUsingHeaders` extracted to `ImageLoader` and then we can make one final decision which set of changes would work best for us

That sounds perfect, thanks so much @kidroca 👍